test/connectivity/NetworkController/Program.cs

```diff
@@ -0,0 +1,132 @@
+        const int DockerNetworkInterfaceStartIndex = 0;
+
+        public static int Main() => MainAsync().Result;
+
+        static async Task<int> MainAsync()
+        {
+            var cts = new CancellationTokenSource();
+            AssemblyLoadContext.Default.Unloading += ctx => cts.Cancel();
+            Console.CancelKeyPress += (sender, cpe) => cts.Cancel();
+            var tcs = new TaskCompletionSource<bool>();
+            cts.Token.Register(s => ((TaskCompletionSource<bool>)s).SetResult(true), tcs);
+
+            switch (Settings.Current.NetworkControllerMode)
+            {
+                case NetworkControllerMode.Offline:
+                    await StartOfflineMode();
+                    break;
+                case NetworkControllerMode.Satellite:
+                    await StartSatelliteMode();
+                    break;
+            }
+
+            await tcs.Task;
+            return 0;
+        }
+
+        static Task StartSatelliteMode() => throw new NotImplementedException();
+
+        static async Task StartOfflineMode()
+        {
+            var delay = Settings.Current.StartAfter;
+            var networkInterfaceName = GetDockerInterfaceName();
+            IOfflineController offlineController;
+
+            if (networkInterfaceName.HasValue)
+            {
+                await networkInterfaceName.ForEachAsync(
+                    async name =>
+                    {
+                        offlineController = new NetworkInterfaceOfflineController(name);
+
+                        foreach (Frequency item in Settings.Current.Frequency)
+                        {
+                            await RunScheduledTask(delay, offlineController, item.OfflineFrequency, item.OnlineFrequency, item.RunsCount);
+                            delay = item.OnlineFrequency;
+                        }
+
+                        offlineController = new FirewallOfflineController(name);
+                        foreach (Frequency item in Settings.Current.Frequency)
+                        {
+                            await RunScheduledTask(delay, offlineController, item.OfflineFrequency, item.OnlineFrequency, item.RunsCount);
+                            delay = item.OnlineFrequency;
+                        }
+                    });
+            }
+            else
+            {
+                Log.LogError($"No network interface found for docker network {Settings.Current.NetworkId}");
+            }
+        }
+
+        static Task RunScheduledTask(TimeSpan delay, IOfflineController offlineController, TimeSpan offlineFreq, TimeSpan onlineFrequency, int runs)
+        {
+            Log.LogInformation(
+                $"Schedule task with {offlineController.Description} to start after {delay} Offline frequency {offlineFreq} Online frequency {onlineFrequency} Run times {runs}");
+
+            INetworkStatusReporter reporter = new NetworkReporter();
+
+            return new ScheduledTask(
```

Comment:
I think you need to report network status online first before scheduled task.

Reply:
I added a method to check the status and remove all rules that were previously set in case the test fails while it was set to offline.

---
In reply to: [353901043](https://github.com/Azure/iotedge/pull/2040#discussion_r353901043) [](ancestors = 353901043)